docs/architecture/adr-003-stateful-precompiles.md

```diff
@@ -0,0 +1,192 @@
+# ADR 003: Stateful Precompiled Contracts
+
+## Changelog
+
+- 2022-06-16: first draft
+
+## Status
+
+DRAFT
+
+## Abstract
+
+Support stateful precompiled contracts to improve inter-operabilities between EVM smart contracts and native functionalities.
+
+## Context
+
+We need inter-operabilities to allow EVM smart contracts to access native functionalities, like manage native tokens through bank module, send/receive IBC messages through IBC modules.
+
+The EVM hooks solution is not ideal, because the message processing is asynchronously, so the caller can't get the return value.
+
+Precompiled contract can provide a much better interface for smart contract developers. But the default precompiled contract implementation in go-ethereum is stateless, we need to do some patches to support stateful ones properly.
+
+## Decision
+
+### Interface Changes
+
+Change the `PrecompiledContract` interface like this (need to patch `go-ethereum`):
+
+```
+ type PrecompiledContract interface {
+  	 RequiredGas(input []byte) uint64
+-  	 Run(input []byte) ([]byte, error)
++  	 Run(evm *vm.EVM, input []byte, caller common.Address, value *big.Int, readonly bool) ([]byte, error)
+ }
+```
+
+There are extra parameters passed to the precompiled contract:
+
+- `caller`: aka. `msg.sender`.
+- `value`: aka. `msg.value`.
+- `readonly`: it's set to `true` for both `staticcall` and `delegatecall`, in these call types, the callee contract is not supposed to modify states. A stateful contract normally should just fail if it's `true`.
```

Comment:
You can leave the `staticcall` and `delegatecall` as is and remove this field altogether. These calls shouldn't run stateful precompiles

Reply:
But with this parameter, the precompile can return an error when it don't support those calls.
For read-only precompiles, it can support these calls.

Reply:
+1 to @yihuang

Reply:
> the precompile can return an error when it don't support those calls.

Yup, the risk here is that the implementer of the precompile misses this check, which would enable a critical vulnerability like [Moonbeam's](https://medium.com/immunefi/moonbeam-missing-call-check-bugfix-review-6279d609bdc5) and [Aurora's](https://medium.com/immunefi/aurora-infinite-spend-bugfix-review-6m-payout-e635d24273d).  

Wondering if there are other alternatives to adding this as an func argument